stan/math/prim/prob/hmm_marginal_lpdf.hpp

```diff
@@ -0,0 +1,187 @@
+#ifndef STAN_MATH_REV_FUN_HMM_MARGINAL_LPDF_HPP
+#define STAN_MATH_REV_FUN_HMM_MARGINAL_LPDF_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/row.hpp>
+#include <stan/math/prim/fun/col.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
+#include <stan/math/prim/fun/exp.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
+#include <stan/math/prim/core.hpp>
+#include <vector>
+#include <iostream>
+
+namespace stan {
+namespace math {
+
+/**
+ * For a Hidden Markov Model with observation y, hidden state x,
+ * and parameters theta, return the log marginal density, log
+ * p(y | theta). In this setting, the hidden states are discrete
+ * and take values over the finite space {1, ..., K}.
+ * The marginal lpdf is obtained via a forward pass, and
+ * the derivative is calculated with an adjoint method,
+ * see (Betancourt, Margossian, & Leos-Barajas, 2020).
+ *
+ * @tparam T_omega type of the log likelihood matrix
+ * @tparam T_Gamma type of the transition matrix
+ * @tparam T_rho type of the initial guess vector
+ *
```

Comment:
I figured it would make the code more readable to explain what each element is. If you insist, I'll make it more concise and relegate to the doc, but I think it's not too long as it is.

Reply:
Check out the doxygen that gets printed.  If it's like Javadoc, only the first sentence is rendered as argument documentation.  Also, the first sentence of the whole function documentation will be rendered as the short form in the indexes.

I think we should follow the convention in things like the C++ standard library doc.  The arguments have very terse doc and the explanation is in the body.  The comments on parameters is more to remind users what the parameter means.

Reply:
Ok, that makes sense.